### ISSUE TITLE
Default to current date in jdate no longer works.

### ISSUE THREAD

**Issue by joomlahenk** [joomlahenk/fabrik#377]
Default to current date in jdate no longer works.
If I revert in jdate.php on line 1097: $isNullDate = $nullDate == $value || $shortNullDate == $value || $value == null;
back to: $isNullDate = $nullDate == $value || $shortNullDate == $value;
then the issue is solved.

I could not find the PR that changed it, nor the issue why we changed it.

**Comment by trob** [comment]
You can see  any change in the code with "Blame".
![grafik](https://user-images.githubusercontent.com/1045355/196398539-00d4fa0d-606b-4144-86ad-b793ac5f7505.png)

@skurvish : what was the issue?

**Comment by joomlahenk** [comment]
Found: PR#340

**Comment by joomlahenk** [comment]
"Blame" howto ??? In the PR search box ??? On command line ??

**Comment by trob** [comment]
Open a file
![grafik](https://user-images.githubusercontent.com/1045355/196401169-328c736e-9ff7-41f4-a6aa-7489b126e423.png)

**Comment by joomlahenk** [comment]
@trob: Got it, thanks

**Comment by skurvish** [comment]
The issue was that if $value was null we got a deprecated error. Don't remember where. @joomlahenk after changing ti back are you getting a deprecated error now?

**Comment by joomlahenk** [comment]
With $isNullDate    = $nullDate == $value || $shortNullDate == $value || $value == null;
I get 1 warning:
Warning: Array to string conversion in /var/www/html/j4test/libraries/fabrik/fabrik/Document/Renderer/Partial/HeadRenderer.php on line 248

With $isNullDate    = $nullDate == $value || $shortNullDate == $value;
I get 4 warnings:
Warning: Array to string conversion in /var/www/html/j4test/libraries/fabrik/fabrik/Document/Renderer/Partial/HeadRenderer.php on line 248
Deprecated: DateTime::__construct(): Passing null to parameter #1 ($datetime) of type string is deprecated in /var/www/html/j4test/libraries/src/Date/Date.php on line 124
Deprecated: Function strftime() is deprecated in /var/www/html/j4test/libraries/src/HTML/HTMLHelper.php on line 1076
Deprecated: Function strftime() is deprecated in /var/www/html/j4test/libraries/src/HTML/HTMLHelper.php on line 1076

**Comment by skurvish** [comment]
I think the problem may be with the change we made to the sql tables for date-time columns, as in Null is allowed and Null is the default. The code as was previously written allowed a null value to pass through with a call to the getDate function which popped up the Deprecated warning in Date @124. The change I made was to treat a null value as a null date and return. In the older sql code it would set the date-time column to current_timestamp but not any more. 

The code in jdate will need to be looked at more closely to determine what the appropriate action should be. If neither of you have time right now I will look at it once I finish a couple of other items I am working on.

**Comment by ontargetWeb** [comment]
To get around this issue I have set Eval to Yes on the jDate element with:
date_default_timezone_set('Europe/Dublin');
$date = date('Y-m-d H:i:s');
return $date;

**Comment by skurvish** [comment]
Yes that will work for now but should not be needed. As indicated above the problem is related to the change we made to the datetime column defaults. I am looking at how best to solve this. I need to deal with timezones and whether the datetime is stored as UTC or not.

**Comment by joomlahenk** [comment]
Strange if I set a print_r("Date ".$date);exit; on line 123 in libraries/src/Date/Date.php just before opening the form, I get "now" as expected not NULL.

**Comment by joomlahenk** [comment]
The deprecations are on PHP8.1 not on PHP8.0.

Joomla also sets NULL as default for type datetime in the db; see the content table. If I create an article then NULL is stored in some datetime columns. Maybe they never do a getDate on these columns and never get the DateTime::__construct() deprecation or they store a valid date before that?
So the change you made to the sql tables for date-time columns is correct and the same as Joomla does.

What I observed:
The DateTime::__construct() deprecation warning is only on opening a form to add a new record, not on edit.
On edit the datetime value can be a valid date or NULL.
The strftime() deprecation warning is on opening a form to add a new record and on edit only if the datetime is valid (not NULL)
The strftime() deprecation warning does not occur on opening a form to edit if the datetime is NULL

**Comment by skurvish** [comment]
OK, I see what the problem is. It has started to happen for 2 reasons, first the change to the sql to allow null and default null, and the second a change to fix a deprecated warning. I should be able to resolve this today. I will also look at the deprecated warnins listed above.

**Comment by skurvish** [comment]
I think PR #402 will finally fix this issue. Please note: You must either define a default on the options tab, or set either or both of default to current date or always return current date to YES. If none of these options are set then you have not told jdate how to default the date/time and if the user does not select a value (and there are not validations requiring them to) the a null value will be stored in the database.

Question, does this need to be documented somewhere? I am not sure if this was the behaviour before or not.

**Comment by joomlahenk** [comment]
I think the behavior is as it should be, if you entered nothing null is stored is ok.

Deprecated: DateTime::__construct() is solved now but instead we have a different one:
Deprecated: urldecode(): Passing null to parameter #1 ($string) of type string is deprecated in /var/www/html/j4test/plugins/fabrik_element/jdate/jdate.php on line 596

strftime() deprecation warning is like before, but that's a Joomla issue.

**Comment by joomlahenk** [comment]
This is now solved with PR #404

**Comment by joomlahenk** [comment]
Back again after PR#436. 
Deprecated: DateTime::__construct(): Passing null to parameter #1 ($datetime) of type string is deprecated in /var/www/html/joomla4upd/libraries/src/Date/Date.php on line 124
